File: UniCalib/unicalib/extrinsic/coarse/feature_matching.py
"""
Camera-Camera 外参粗标定
基于特征匹配 + 本质矩阵分解 + RANSAC
支持普通相机和鱼眼相机
"""
from __future__ import annotations
import logging
from typing import Dict, List, Optional, Tuple

import cv2
import numpy as np
from scipy.spatial.transform import Rotation

from ...core.calib_result import CalibResult, CameraIntrinsic, FisheyeIntrinsic
from ...core.sensor_config import SensorConfig, SensorType
from ...core.data_manager import DataManager

logger = logging.getLogger(__name__)


class FeatureMatchingCalibrator:
    """
    Camera-Camera 外参粗标定器。

    流程:
      1. 从同步图像对中提取特征点 (SIFT / ORB)
      2. 特征匹配 + RANSAC 过滤
      3. 基于本质矩阵分解 R, t
      4. 多帧结果鲁棒集成

    注意: 仅双目可恢复尺度，纯旋转相机对需 LiDAR 约束才能获得平移尺度。
    """

    def __init__(self, system_config: dict):
        cfg = (system_config.get("calibration", {})
               .get("stage2_coarse", {})
               .get("camera_camera", {}))
        self.feature_type = cfg.get("feature_matcher", "sift")  # sift / orb
        self.min_matches = cfg.get("min_matches", 30)
        self.max_frames = cfg.get("max_frames", 50)

    def calibrate(
        self,
        data_mgr: DataManager,
        sensor_a: SensorConfig,
        sensor_b: SensorConfig,
        intrinsics: Dict,
    ) -> CalibResult:
        """执行 Camera-Camera 外参粗标定。"""

        intr_a = intrinsics.get(sensor_a.sensor_id)
        intr_b = intrinsics.get(sensor_b.sensor_id)

        if intr_a is None or intr_b is None:
            logger.warning(f"Missing intrinsics for camera pair "
                           f"{sensor_a.sensor_id}-{sensor_b.sensor_id}")
            return self._identity_result(sensor_a, sensor_b)

        # 收集同步帧对
        frame_pairs = list(data_mgr.iter_synced_frames(
            sensor_a.topic, sensor_b.topic,
            sync_threshold_ms=100.0,
            max_pairs=self.max_frames
        ))

        if len(frame_pairs) < 3:
            logger.warning(f"  Insufficient synced frames ({len(frame_pairs)}) "
                           f"for {sensor_a.sensor_id}-{sensor_b.sensor_id}")
            return self._identity_result(sensor_a, sensor_b)

        logger.info(f"  Feature matching on {len(frame_pairs)} frame pairs...")

        # 提取匹配
        all_pts_a, all_pts_b = [], []
        for _, img_a, img_b in frame_pairs:
            pts_a, pts_b = self._match_features(
                img_a, img_b, intr_a, intr_b,
                sensor_a.is_fisheye(), sensor_b.is_fisheye()
            )
            if pts_a is not None and len(pts_a) >= self.min_matches:
                all_pts_a.append(pts_a)
                all_pts_b.append(pts_b)

        if not all_pts_a:
            logger.warning("  No valid matches found.")
            return self._identity_result(sensor_a, sensor_b)

        pts_a_all = np.vstack(all_pts_a)
        pts_b_all = np.vstack(all_pts_b)
        logger.info(f"  Total matched points: {len(pts_a_all)}")

        # 估计本质矩阵并分解
        K_a = self._get_pinhole_K(intr_a)
        K_b = self._get_pinhole_K(intr_b)

        E, mask_E = cv2.findEssentialMat(
            pts_a_all, pts_b_all, K_a,
            method=cv2.RANSAC, prob=0.999, threshold=1.0
        )
        if E is None:
            return self._identity_result(sensor_a, sensor_b)

        _, R, t, mask_pose = cv2.recoverPose(
            E, pts_a_all, pts_b_all, K_a, mask=mask_E)

        inlier_ratio = float(np.sum(mask_pose > 0)) / len(pts_a_all)
        logger.info(f"  [{sensor_a.sensor_id}-{sensor_b.sensor_id}] "
                    f"inliers={inlier_ratio:.1%}  "
                    f"R_euler={Rotation.from_matrix(R).as_euler('zyx', degrees=True)}")

        return CalibResult(
            pair=(sensor_a.sensor_id, sensor_b.sensor_id),
            rotation=R,
            translation=t.flatten(),
            confidence=inlier_ratio,
            method_used=f"feature_matching_{self.feature_type}",
        )

    # ------------------------------------------------------------------
    # 特征提取与匹配
    # ------------------------------------------------------------------

    def _match_features(
        self,
        img_a: np.ndarray, img_b: np.ndarray,
        intr_a, intr_b,
        fisheye_a: bool, fisheye_b: bool,
    ) -> Tuple[Optional[np.ndarray], Optional[np.ndarray]]:
        """提取特征点并匹配，返回对应点对。"""

        # 先去畸变
        gray_a = cv2.cvtColor(img_a, cv2.COLOR_BGR2GRAY)
        gray_b = cv2.cvtColor(img_b, cv2.COLOR_BGR2GRAY)

        gray_a = self._undistort(gray_a, intr_a, fisheye_a)
        gray_b = self._undistort(gray_b, intr_b, fisheye_b)

        # 提取关键点
        detector = self._get_detector()
        kp_a, desc_a = detector.detectAndCompute(gray_a, None)
        kp_b, desc_b = detector.detectAndCompute(gray_b, None)

        if desc_a is None or desc_b is None or len(kp_a) < 10 or len(kp_b) < 10:
            return None, None

        # 匹配
        matches = self._match(desc_a, desc_b)
        if len(matches) < self.min_matches:
            return None, None

        pts_a = np.float32([kp_a[m.queryIdx].pt for m in matches])
        pts_b = np.float32([kp_b[m.trainIdx].pt for m in matches])

        return pts_a, pts_b

    def _get_detector(self):
        """创建特征检测器。"""
        if self.feature_type == "sift":
            try:
                return cv2.SIFT_create(nfeatures=500)
            except AttributeError:
                return cv2.xfeatures2d.SIFT_create(500)
        elif self.feature_type == "orb":
            return cv2.ORB_create(nfeatures=500)
        else:
            return cv2.SIFT_create(nfeatures=500)

    def _match(self, desc_a, desc_b) -> List:
        """BFMatcher + Lowe 比率测试。"""
        norm = cv2.NORM_L2 if self.feature_type == "sift" else cv2.NORM_HAMMING
        matcher = cv2.BFMatcher(norm, crossCheck=False)
        raw = matcher.knnMatch(desc_a, desc_b, k=2)
        good = []
        for pair in raw:
            if len(pair) == 2:
                m, n = pair
                if m.distance < 0.75 * n.distance:
                    good.append(m)
        return good

    def _undistort(self, gray: np.ndarray, intrinsic, is_fisheye: bool) -> np.ndarray:
        """对图像进行去畸变。"""
        if isinstance(intrinsic, CameraIntrinsic):
            K = intrinsic.K
            D = intrinsic.dist_coeffs
            if is_fisheye:
                h, w = gray.shape[:2]
                K_new = cv2.fisheye.estimateNewCameraMatrixForUndistortRectify(
                    K, D[:4].reshape(4, 1), (w, h), np.eye(3))
                gray = cv2.fisheye.undistortImage(gray, K, D[:4].reshape(4, 1), K_new)
            else:
                gray = cv2.undistort(gray, K, D)
        return gray

    def _get_pinhole_K(self, intrinsic) -> np.ndarray:
        """从任意内参类型获取 3x3 K 矩阵。"""
        if isinstance(intrinsic, CameraIntrinsic):
            return intrinsic.K
        elif isinstance(intrinsic, FisheyeIntrinsic):
            p = intrinsic.params
            return np.array([
                [p.get("fx", 500), 0, p.get("cx", 320)],
                [0, p.get("fy", 500), p.get("cy", 240)],
                [0, 0, 1],
            ], dtype=np.float64)
        else:
            return np.eye(3) * 500.0

    # ------------------------------------------------------------------
    # 辅助
    # ------------------------------------------------------------------

    def _identity_result(self, sa: SensorConfig, sb: SensorConfig) -> CalibResult:
        return CalibResult(
            pair=(sa.sensor_id, sb.sensor_id),
            rotation=np.eye(3),
            translation=np.zeros(3),
            confidence=0.0,
            method_used="identity_fallback",
        )

File: unicalib_C_plus_plus/src/feature_matching.cpp
/**
 * UniCalib C++ — Camera-Camera 特征匹配粗外参（OpenCV）
 */
#include "unicalib/feature_matching.hpp"
#include "unicalib/logger.hpp"
#include <vector>
#include <cmath>
#include <algorithm>

#if defined(UNICALIB_USE_OPENCV) && UNICALIB_USE_OPENCV
#include <opencv2/calib3d.hpp>
#include <opencv2/features2d.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#endif

namespace unicalib {

namespace {

constexpr double kSyncThresholdSec = 0.1;

#if defined(UNICALIB_USE_OPENCV) && UNICALIB_USE_OPENCV
bool get_K_from_intrinsic(const IntrinsicResultHolder& h, cv::Mat& K) {
  if (h.camera_pinhole) {
    K = (cv::Mat_<double>(3, 3) <<
         h.camera_pinhole->K(0, 0), 0, h.camera_pinhole->K(0, 2),
         0, h.camera_pinhole->K(1, 1), h.camera_pinhole->K(1, 2),
         0, 0, 1);
    return true;
  }
  if (h.camera_fisheye) {
    auto it_fx = h.camera_fisheye->params.find("fx");
    auto it_fy = h.camera_fisheye->params.find("fy");
    auto it_cx = h.camera_fisheye->params.find("cx");
    auto it_cy = h.camera_fisheye->params.find("cy");
    if (it_fx == h.camera_fisheye->params.end() || it_fy == h.camera_fisheye->params.end()) return false;
    double fx = it_fx->second, fy = it_fy->second;
    double cx = it_cx != h.camera_fisheye->params.end() ? it_cx->second : 0;
    double cy = it_cy != h.camera_fisheye->params.end() ? it_cy->second : 0;
    K = (cv::Mat_<double>(3, 3) << fx, 0, cx, 0, fy, cy, 0, 0, 1);
    return true;
  }
  return false;
}
#endif

}  // namespace

std::optional<CalibResult> run_feature_matching_coarse(
  DataManager& data_mgr,
  const SensorConfig& sensor_a,
  const SensorConfig& sensor_b,
  const std::unordered_map<std::string, IntrinsicResultHolder>& intrinsics,
  int max_frames,
  int min_matches) {
#if defined(UNICALIB_USE_OPENCV) && UNICALIB_USE_OPENCV
  auto it_a = intrinsics.find(sensor_a.sensor_id);
  auto it_b = intrinsics.find(sensor_b.sensor_id);
  if (it_a == intrinsics.end() || it_b == intrinsics.end()) return std::nullopt;

  cv::Mat K_a, K_b;
  if (!get_K_from_intrinsic(it_a->second, K_a) || !get_K_from_intrinsic(it_b->second, K_b))
    return std::nullopt;

  std::vector<std::pair<double, std::string>> paths_a, paths_b;
  data_mgr.iter_image_paths(sensor_a.topic, [&](double ts, const std::string& path) {
    paths_a.emplace_back(ts, path);
  }, max_frames * 2, 1);
  data_mgr.iter_image_paths(sensor_b.topic, [&](double ts, const std::string& path) {
    paths_b.emplace_back(ts, path);
  }, max_frames * 2, 1);
  if (paths_a.empty() || paths_b.empty()) return std::nullopt;

  std::sort(paths_a.begin(), paths_a.end());
  std::sort(paths_b.begin(), paths_b.end());

  std::vector<cv::Point2f> all_pts_a, all_pts_b;
  int paired = 0;
  for (const auto& [ts_a, path_a] : paths_a) {
    if (paired >= max_frames) break;
    auto it = std::lower_bound(paths_b.begin(), paths_b.end(), ts_a - kSyncThresholdSec,
        [](const std::pair<double, std::string>& x, double t) { return x.first < t; });
    if (it == paths_b.end() || std::abs(it->first - ts_a) > kSyncThresholdSec) continue;

    cv::Mat img_a = cv::imread(path_a);
    cv::Mat img_b = cv::imread(it->second);
    if (img_a.empty() || img_b.empty()) continue;

    cv::Mat gray_a, gray_b;
    cv::cvtColor(img_a, gray_a, cv::COLOR_BGR2GRAY);
    cv::cvtColor(img_b, gray_b, cv::COLOR_BGR2GRAY);

    cv::Ptr<cv::SIFT> sift = cv::SIFT::create(200);
    std::vector<cv::KeyPoint> kp_a, kp_b;
    cv::Mat desc_a, desc_b;
    sift->detectAndCompute(gray_a, cv::noArray(), kp_a, desc_a);
    sift->detectAndCompute(gray_b, cv::noArray(), kp_b, desc_b);
    if (desc_a.empty() || desc_b.empty()) continue;

    cv::BFMatcher matcher(cv::NORM_L2);
    std::vector<cv::DMatch> matches;
    matcher.match(desc_a, desc_b, matches);
    if (matches.size() < static_cast<size_t>(min_matches)) continue;

    std::vector<cv::Point2f> pts_a, pts_b;
    for (const auto& m : matches) {
      pts_a.push_back(kp_a[m.queryIdx].pt);
      pts_b.push_back(kp_b[m.trainIdx].pt);
    }

    cv::Mat E, R, t;
    cv::Mat mask;
    E = cv::findEssentialMat(pts_a, pts_b, K_a, cv::RANSAC, 0.999, 1.0, mask);
    if (E.empty()) continue;
    int inliers = cv::recoverPose(E, pts_a, pts_b, K_a, R, t, mask);
    if (inliers < min_matches) continue;

    for (int i = 0; i < mask.rows; ++i) {
      if (mask.at<unsigned char>(i)) {
        all_pts_a.push_back(pts_a[i]);
        all_pts_b.push_back(pts_b[i]);
      }
    }
    ++paired;
  }

  if (all_pts_a.size() < static_cast<size_t>(min_matches)) {
    LOG_WARNING("Feature matching: insufficient inliers");
    return std::nullopt;
  }

  cv::Mat E2, R, t, mask2;
  E2 = cv::findEssentialMat(all_pts_a, all_pts_b, K_a, cv::RANSAC, 0.999, 1.0, mask2);
  if (E2.empty()) return std::nullopt;
  cv::recoverPose(E2, all_pts_a, all_pts_b, K_a, R, t, mask2);

  CalibResult out;
  out.pair = {sensor_a.sensor_id, sensor_b.sensor_id};
  out.method_used = "feature_matching";
  out.confidence = 0.6;
  for (int i = 0; i < 3; ++i)
    for (int j = 0; j < 3; ++j)
      out.rotation(i, j) = R.at<double>(i, j);
  out.translation(0) = t.at<double>(0);
  out.translation(1) = t.at<double>(1);
  out.translation(2) = t.at<double>(2);
  double n = out.translation.norm();
  if (n > 1e-6) out.translation /= n;
  LOG_INFO("Feature matching coarse OK: " + sensor_a.sensor_id + " -> " + sensor_b.sensor_id);
  return out;
#else
  (void)data_mgr;
  (void)sensor_a;
  (void)sensor_b;
  (void)intrinsics;
  (void)max_frames;
  (void)min_matches;
  LOG_WARNING("Feature matching requires OpenCV");
  return std::nullopt;
#endif
}

}  // namespace unicalib
